### PR TITLE
Request.destination value tests

### PR DESCRIPTION
**Pull request by yoavweiss**
Solves #8810.

This is not yet a complete test suite, but as I'm having trouble running this locally, I want to see these pass on the CI bots before creating more test cases.

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/22201)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/322432618)

**Comment by yoavweiss** [comment]
This is now ready to review

File: fetch/api/request/destination/fetch-destination-style.https.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      .then(t.step_func(function(frame) {
+          var link = frame.contentWindow.document.createElement("link");
+          link.rel = "stylesheet";
+          link.addEventListener("load", t.step_func_done(function() { registration.unregister(); }));
+          link.addEventListener("error", t.step_func_done(function() {
+              registration.unregister();
+              assert_unreached("Fetch destination preload errored");
+          }));
+          link.href = "resources/dummy?dest=style";
+          frame.contentWindow.document.body.appendChild(script);
```
Comment (jstenback) [inline fetch/api/request/destination/fetch-destination-style.https.html]: I'm assuming you meant to append 'link' here, not 'script'?

Comment (yoavweiss) [reply fetch/api/request/destination/fetch-destination-style.https.html]: Yeah, in the latest version of the tests, it's all just `node`

File: fetch/api/request/destination/fetch-destination-script.https.html [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+          node.addEventListener("load", t.step_func_done(function() { registration.unregister(); }));
+          node.addEventListener("error", t.step_func_done(function() {
+              registration.unregister();
+              assert_unreached("Fetch destination script errored");
+          }));
+          node.src = "dummy?dest=script";
+          frame.contentWindow.document.body.appendChild(node);
+        }))
+      .catch(unreached_rejection(t));
+}, 'Fetch destination script');
```
Comment (wanderview) [inline fetch/api/request/destination/fetch-destination-script.https.html]: The "script" destination can also be produced by worker `importScripts()`.  We should have a test for that as well.

File: fetch/api/request/destination/fetch-destination-style.https.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+          node.addEventListener("load", t.step_func_done(function() { registration.unregister(); }));
+          node.addEventListener("error", t.step_func_done(function() {
+              registration.unregister();
+              assert_unreached("Fetch destination style errored");
+          }));
+          node.href = "dummy?dest=style";
+          frame.contentWindow.document.body.appendChild(node);
+        }))
+      .catch(unreached_rejection(t));
+}, 'Fetch destination style');
```
Comment (wanderview) [inline fetch/api/request/destination/fetch-destination-style.https.html]: We should have a test for CSS `@import` which can produce a "style" destination as well.

File: fetch/api/request/destination/fetch-destination-image.https.html [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+          var node = frame.contentWindow.document.createElement("img");
+          node.addEventListener("load", t.step_func_done(function() { registration.unregister(); }));
+          node.addEventListener("error", t.step_func_done(function() {
+              registration.unregister();
+              assert_unreached("Fetch destination image errored");
+          }));
+          node.src = "dummy.png?dest=image";
+        }))
+      .catch(unreached_rejection(t));
+}, 'Fetch destination image');
```
[the remark this answers is not in the document]
Comment (wanderview) [reply fetch/api/request/destination/fetch-destination-image.https.html]: Also imageset and picture elements.

**Review by wanderview — CHANGES_REQUESTED** [review]
This is a good start, but its missing a lot of coverage.  The big challenge with a feature like destination is trying to make sure you cover all the various network requests the browser can generate.  I commented on some that are missing below, but we also need tests for these destination types:

* report
* document
* font
* sharedworker
* worker
* track
* manifest
* xslt

Not sure if we need "embed" or "object" since in theory those systems are being deprecated, I think.

Basically, we should try to test as many of the items in the table in the spec here:

https://fetch.spec.whatwg.org/#concept-request-destination

File: fetch/api/request/destination/fetch-destination-audio.https.html [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="../../../../service-workers/service-worker/resources/test-helpers.sub.js"></script>
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/fetch-destination-audio.https.html]: I feel like these tests are better here, as they test the Fetch API *using Service Workers*. So I prefer to keep them here. Let me know if you vehemently disagree :)

File: fetch/api/request/destination/fetch-destination-fetch.https.html [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+          return wait_for_state(t, r.installing, 'activated');
+        }))
+      .then(t.step_func(function() {
+          return with_iframe(scope);
+        }))
+      .then(t.step_func(function(frame) {
+          frame.contentWindow.registration = registration;
+          var node = frame.contentWindow.document.createElement("script");
+          var text = frame.contentWindow.document.createTextNode(
+              'fetch("dummy?dest=").then(function(response) { ' +
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/fetch-destination-fetch.https.html]: Trying that, but I'm getting a "TypeError: Failed to fetch"...

[the remark this answers is not in the document]
Comment (wanderview) [reply fetch/api/request/destination/fetch-destination-fetch.https.html]: That doesn't sound right.  We do that all the time in other service worker tests:

https://searchfox.org/mozilla-central/source/testing/web-platform/tests/service-workers/service-worker/fetch-event-after-navigation-within-page.https.html#23

File: fetch/api/request/destination/resources/fetch-destination-worker.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+self.addEventListener('fetch', function(event) {
+    if (event.request.url.indexOf('dummy') != -1) {
+        let destination = new URL(event.request.url).searchParams.get("dest");
+        if (event.request.destination == destination)
+            event.respondWith(fetch(event.request));
+        else
+            event.respondWith(Response.error());
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/resources/fetch-destination-worker.js]: sure

File: fetch/api/request/destination/resources/fetch-destination-worker.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+self.addEventListener('fetch', function(event) {
+    if (event.request.url.indexOf('dummy') != -1) {
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/resources/fetch-destination-worker.js]: indeed

File: fetch/api/request/destination/fetch-destination-fetch.https.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/fetch-destination-fetch.https.html]: Got it working

File: fetch/api/request/destination/fetch-destination-audio.https.html [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        }))
+      .then(t.step_func(function() {
+          return with_iframe(scope);
+        }))
+      .then(t.step_func(function(frame) {
+          var audioURL = getAudioURI("dummy_audio");
+          var node = frame.contentWindow.document.createElement("audio");
+          node.addEventListener("loadstart", t.step_func_done(function() { registration.unregister(); }));
+          node.addEventListener("error", t.step_func_done(function() {
+              registration.unregister();
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/fetch-destination-audio.https.html]: ok

File: fetch/api/request/destination/fetch-destination-audio.https.html [hunk 10]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="../../../../service-workers/service-worker/resources/test-helpers.sub.js"></script>
+<script src="/common/media.js"></script>
+<script>
+async_test(function(t) {
```
[the remark this answers is not in the document]
Comment (yoavweiss) [reply fetch/api/request/destination/fetch-destination-audio.https.html]: refactored

**Comment by yoavweiss** [comment]
@wanderview I addressed your comments on the existing tests. If possible I prefer to land this subset soon, and continue working on adding the missing destinations (that you rightfully pointed out) in separate patches.

File: fetch/api/request/destination/fetch-destination.https.html [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+}, 'HTMLImageElement fetches with an "image" Request.destination');
+
+// fetch() - empty string destination
+promise_test(async t => {
+  let response = await frame.contentWindow.fetch("dummy?dest=");
+  assert_true(response.ok);
+}, 'fetch() fetches with an empty string Request.destination');
+
+// HTMLAudioElement - audio destination
+promise_test(async t => {
```
Comment (wanderview) [inline fetch/api/request/destination/fetch-destination.https.html]: nit: Thanks for trying to use more async functions as I requested.  For the cases where you have a single `await new Promise()` in the function you could probably just switch to `promise_test(new Promise()` instead.  What you have here works, though, and more just mentioning this in case you felt forced to use the await per my previous comment.

**Review by wanderview — APPROVED** [review]
Thanks!  LGTM, but as we discussed offline it would be great to have more test coverage before shipping the feature.

**Comment by yoavweiss** [comment]
Thanks for reviewing! :)
